File: ios/Classes/PrinterConnectFilterUtil.swift
import CoreBluetooth
import Foundation

/// 扫描过滤器工具类。
///
/// 支持三种自定义过滤条件：
/// - 设备名称前缀匹配
/// - 服务 UUID 匹配
/// - 制造商数据（Company ID + Payload 前缀 + 掩码）匹配
/// 多个条件之间采用 OR 逻辑（任一条件满足即通过）。
class PrinterConnectFilterUtil {

    /// 自定义扫描过滤器配置（来自 Dart 层的 UniversalScanFilter）
    var scanFilter: UniversalScanFilter?
    /// 预解析的服务 UUID 列表（用于快速匹配）
    var scanFilterServicesUUID: [CBUUID] = []

    /// 判断外设是否通过过滤器。
    /// 采用 OR 逻辑：名称前缀匹配 || 服务 UUID 匹配 || 制造商数据匹配。
    /// 如果没有配置任何过滤条件，默认通过。
    func filterDevice(name: String?, manufacturerData: UniversalManufacturerData?, services: [CBUUID]?) -> Bool {
        guard let filter = scanFilter else {
            return true
        }
        let hasNamePrefixFilter = !filter.withNamePrefix.isEmpty
        let hasServiceFilter = !filter.withServices.isEmpty
        let hasManufacturerDataFilter = !filter.withManufacturerData.isEmpty

        if !hasNamePrefixFilter && !hasServiceFilter && !hasManufacturerDataFilter {
            return true
        }

        return hasNamePrefixFilter && isNameMatchingFilters(filter: filter, name: name) ||
            hasServiceFilter && isServicesMatchingFilters(services: services) ||
            hasManufacturerDataFilter && isManufacturerDataMatchingFilters(scanFilter: filter, msd: manufacturerData)
    }

    /// 检查设备名称是否匹配名称前缀过滤条件。
    /// 只要设备名称以任意一个配置的前缀开头即通过。
    func isNameMatchingFilters(filter: UniversalScanFilter, name: String?) -> Bool {
        let prefixFilters = filter.withNamePrefix.compactMap { $0 }.filter { !$0.isEmpty }
        guard !prefixFilters.isEmpty else {
            return true
        }
        guard let name = name, !name.isEmpty else {
            return false
        }
        return prefixFilters.contains { name.hasPrefix($0) }
    }

    /// 检查广播中的服务 UUID 是否与过滤条件有交集。
    /// 使用集合交集判断：只要广播服务中包含任意一个被过滤的服务即通过。
    func isServicesMatchingFilters(services: [CBUUID]?) -> Bool {
        let serviceFilters = Set(scanFilterServicesUUID.compactMap { $0 })
        guard !serviceFilters.isEmpty else {
            return true
        }
        guard let services = services, !services.isEmpty else {
            return false
        }
        return !Set(services).isDisjoint(with: serviceFilters)
    }

    /// 检查制造商数据是否匹配过滤条件。
    /// 依次比较每个过滤规则：先匹配 Company Identifier，再通过 findData 进行 Payload 前缀+掩码匹配。
    func isManufacturerDataMatchingFilters(scanFilter: UniversalScanFilter, msd: UniversalManufacturerData?) -> Bool {
        let filters = scanFilter.withManufacturerData.compactMap { $0 }
        if filters.isEmpty {
            return true
        }
        guard let msd = msd else {
            return false
        }
        for filter in filters {
            let companyIdentifier: Int64 = filter.companyIdentifier
            if msd.companyIdentifier == companyIdentifier && findData(find: filter.payloadPrefix?.toData(), inData: msd.data.toData(), usingMask: filter.payloadMask?.toData()) {
                return true
            }
        }
        return false
    }

    /// 在数据中查找匹配的字节序列。
    /// 使用掩码（mask）逐字节进行 AND 比较，实现灵活的字节匹配逻辑。
    /// 如果 find 为 nil 或空，直接返回 true。
    func findData(find: Data?, inData data: Data, usingMask mask: Data?) -> Bool {
        if let find = find {
            let mask = mask ?? Data(repeating: 0xFF, count: find.count)
            guard find.count == mask.count else {
                return false
            }
            for i in 0 ..< find.count {
                if (find[i] & mask[i]) != (data[i] & mask[i]) {
                    return false
                }
            }
        }
        return true
    }
}

// MARK: - UniversalScanFilter 扩展

extension UniversalScanFilter {
    /// 判断是否使用了自定义过滤器（制造商数据过滤或名称前缀过滤）
    /// 如果使用了自定义过滤器，系统级服务过滤将被禁用，由自定义逻辑全权处理
    var usesCustomFilters: Bool {
        return !withManufacturerData.isEmpty || !withNamePrefix.isEmpty
    }
}

File: android/src/main/kotlin/com/example/printer_connect/PrinterConnectFilterUtil.kt
package com.example.printer_connect

import android.annotation.SuppressLint
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.os.ParcelUuid
import android.util.Log
import java.util.UUID
import kotlin.experimental.and

// 过滤器工具类，负责 BLE 扫描结果的过滤逻辑
// 支持三种过滤方式：名称前缀、服务 UUID、制造商数据
// 工作模式分为：自定义过滤（Dart 端过滤）和原生过滤（Android ScanFilter）
object PrinterConnectFilterUtil {

    // 当前设置的通用扫描过滤器（Dart 端传入）
    var scanFilter: UniversalScanFilter? = null
    // 原生过滤器使用的服务 UUID 列表
    var serviceFilterUUIDS: List<UUID> = emptyList()

    // 核心过滤方法，根据名称、制造商数据和服务 UUID 判断设备是否匹配
    // 返回 true 表示设备通过所有过滤条件
    fun filterDevice(
        name: String?,
        manufacturerDataList: List<UniversalManufacturerData>,
        serviceUuids: Array<UUID>,
    ): Boolean {
        val filter = scanFilter ?: return true
        val hasNamePrefixFilter = filter.withNamePrefix.isNotEmpty()
        val hasServiceFilter = filter.withServices.isNotEmpty()
        val hasManufacturerDataFilter = filter.withManufacturerData.isNotEmpty()

        if (!hasNamePrefixFilter &&
            !hasServiceFilter &&
            !hasManufacturerDataFilter
        ) {
            return true
        }

        return hasNamePrefixFilter && isNameMatchingFilters(filter, name) ||
                hasServiceFilter && isServicesMatchingFilters(serviceUuids) ||
                hasManufacturerDataFilter && isManufacturerDataMatchingFilters(
            filter,
            manufacturerDataList
        )
    }

    // 名称前缀匹配：检查设备名称是否以任一过滤前缀开头
    // 若无前缀规则则直接通过；设备名为空则不匹配
    private fun isNameMatchingFilters(scanFilter: UniversalScanFilter, name: String?): Boolean {
        val namePrefixFilter = scanFilter.withNamePrefix
        if (namePrefixFilter.isEmpty()) {
            return true
        }
        if (name.isNullOrEmpty()) {
            return false
        }
        return namePrefixFilter.any { name.startsWith(it) }
    }

    // 服务 UUID 匹配：检查设备广播的服务 UUID 是否包含任一过滤 UUID
    // 若过滤列表为空则直接通过；设备无服务 UUID 则不匹配
    private fun isServicesMatchingFilters(
        serviceUuids: Array<UUID>,
    ): Boolean {
        if (serviceFilterUUIDS.isEmpty()) {
            return true
        }
        if (serviceUuids.isEmpty()) {
            return false
        }
        return serviceFilterUUIDS.any {
            serviceUuids.contains(it)
        }
    }

    // 制造商数据匹配：遍历设备的制造商数据列表，检查是否满足任一过滤条件
    // 匹配规则：公司 ID 一致 且 数据载荷通过位掩码匹配
    private fun isManufacturerDataMatchingFilters(
        scanFilter: UniversalScanFilter,
        manufacturerDataList: List<UniversalManufacturerData>,
    ): Boolean {
        val filters = scanFilter.withManufacturerData
        if (filters.isEmpty()) return true
        if (manufacturerDataList.isEmpty()) return false
        return manufacturerDataList.any { msd ->
            filters.any { filter ->
                msd.companyIdentifier == filter.companyIdentifier &&
                        isDataMatching(filter.payloadPrefix, msd.data, filter.payloadMask)
            }
        }
    }

    // 数据匹配算法：使用位掩码逐字节比对过滤前缀与设备数据
    // 掩码为 null 时默认为全 0xFF（所有位都参与比较）
    // 匹配规则：(filterData[i] and mask[i]) == (deviceData[i] and mask[i])
    private fun isDataMatching(
        filterData: ByteArray?,
        deviceData: ByteArray,
        filterMask: ByteArray?,
    ): Boolean {
        if (filterData == null) return true
        if (filterData.size > deviceData.size) return false
        val mask = filterMask ?: ByteArray(filterData.size) { 0xFF.toByte() }
        if (filterData.size != mask.size) return false
        return filterData.indices.all { i ->
            (filterData[i] and mask[i]) == (deviceData[i] and mask[i])
        }
    }
}

// 顶级扩展函数（放在对象外部，便于以 `filter?.usesCustomFilters()` / `filter.toScanFilters(...)` 的形式直接调用，
// 与参考项目结构保持一致。

// 判断是否需要使用自定义过滤（Dart 端过滤）。
// 选择策略：名称前缀过滤在原生 ScanFilter 中不支持，必须使用自定义过滤；
// 制造商数据和服务 UUID 可通过原生 ScanFilter 实现，无需自定义过滤。
fun UniversalScanFilter?.usesCustomFilters(): Boolean {
    if (this == null) return false
    return withNamePrefix.isNotEmpty()
}

// 将通用扫描过滤器转换为 Android 原生 ScanFilter 列表
// 用于在 startScan 时直接传递给 BluetoothLeScanner 进行原生过滤
// 包含两类原生过滤：服务 UUID 过滤 和 制造商数据过滤（支持 payloadMask 位掩码）
fun UniversalScanFilter.toScanFilters(serviceUuids: List<UUID>): List<ScanFilter> {
    val scanFilters = mutableListOf<ScanFilter>()

    for (service in serviceUuids) {
        try {
            scanFilters.add(
                ScanFilter.Builder().setServiceUuid(ParcelUuid(service)).build()
            )
        } catch (e: Exception) {
            PrinterConnectLogger.logError("Invalid service UUID: $service")
            throw createFlutterError(
                UniversalBleErrorCode.FAILED,
                "Invalid serviceId: $service",
                e.toString(),
            )
        }
    }

    for (manufacturerData in this.withManufacturerData) {
        try {
            val data = manufacturerData.payloadPrefix ?: ByteArray(0)
            val mask = manufacturerData.payloadMask
            if (mask == null) {
                scanFilters.add(
                    ScanFilter.Builder().setManufacturerData(
                        manufacturerData.companyIdentifier.toInt(), data
                    ).build()
                )
            } else {
                scanFilters.add(
                    ScanFilter.Builder().setManufacturerData(
                        manufacturerData.companyIdentifier.toInt(), data, mask
                    ).build()
                )
            }
        } catch (e: Exception) {
            PrinterConnectLogger.logError("Invalid manufacturerData: ${manufacturerData.companyIdentifier}")
            throw createFlutterError(
                UniversalBleErrorCode.FAILED,
                "Invalid manufacturerData: ${manufacturerData.companyIdentifier} ${manufacturerData.payloadPrefix} ${manufacturerData.payloadMask}",
                e.toString(),
            )
        }
    }

    return scanFilters
}

@SuppressLint("MissingPermission")
// 判断扫描结果是否匹配过滤器。
// 使用 resolvedDeviceName（优先广播名称，其次缓存的 device.name）确保过滤使用的名称源
// 与上报给 Dart 层的名称源一致，避免过滤结果与实际结果不匹配。
fun ScanResult.isDeviceMatchingFilter(filter: UniversalScanFilter?): Boolean {
    if (filter == null) return true
    val name = this.resolvedDeviceName
    val manufacturerData = this.manufacturerDataList
    // scanRecord.serviceUuids 为 List<ParcelUuid>，提取 .uuid 转为 Array<UUID> 用于过滤
    val serviceUuids = scanRecord?.serviceUuids?.map { it.uuid }?.toTypedArray() ?: arrayOf()
    return PrinterConnectFilterUtil.filterDevice(name, manufacturerData, serviceUuids)
}
